embedded/__main__.py

```diff
@@ -60,30 +81,162 @@ def handle_gps_polling():
+        :raises AuthError:
+        """
+    async with session.post(URL + "/connect", data=bike.public_key.encode(RawEncoder)) as resp:
+        if resp.status != 200 and resp.reason == "Identity not recognized.":
+            raise AuthError("public key not on server")
+        return await resp.read()
+
+
+def bike_handler(session, bike, signature):
+    async def bike_handler(session, bike: Bike, signed_challenge: bytes):
+        """
+        Opens an authenticated web socket session with the server.
+        :return: None
+        """
+        async with session.ws_connect(URL + "/connect") as socket:
+            await socket.send_bytes(bike.public_key.encode(RawEncoder))
+            await socket.send_bytes(signed_challenge)
+            confirmation = await socket.receive_str()
+            if "fail" in confirmation:
+                raise AuthError(confirmation.split(":")[1])
+            else:
+                logger.info(f"Bike {bike.bid} established connection")
+                await socket.send_json({"locked": bike.locked})
+            bike.socket = socket
+            async for msg in socket:
+                msg: WSMessage = msg
+                logger.info("Message %s", msg)
+                if msg.type == aiohttp.WSMsgType.TEXT:
+                    try:
+                        data = msg.json()
+                    except Exception:
+                        continue
+                    else:
+                        if "method" in data:
+                            await bike.handle_request(data)
+                        else:
+                            await bike.handle_response(data)
+
+
+class AuthError(Exception):
+    pass
+
+
+async def register_bike(session, bike: Bike, master_key: bytes):
+    bike_register = {
+        "public_key": bike.public_key.encode(RawEncoder),
+        "type": BikeType.ROAD,
+        "master_key": master_key
+    }
+    schema = BikeRegisterSchema()
+
+    bike_serialized = schema.dump(bike_register)
+
+    async with session.post(URL, json=bike_serialized) as resp:
+        text = await resp.text()
+        pass
+
+
+async def runner():
+    async with ClientSession() as session:
+        while not QUIT_EVENT.is_set():
+            try:
+                challenge = await create_ticket(session, bike)
+                signature = bike.sign(challenge).signature
+                await bike_handler(session, bike, signature)
+            except AuthError as e:
+                if "public key not on server" in e.args:
+                    await register_bike(session, bike, 0xdeadbeef.to_bytes(4, "big"))
+                else:
+                    logger.error(f"Bike {bike.bid} {e}, quitting...")
+                    return
+            except ClientConnectorError as e:
+                logger.error("Connection lost, retrying..")
+                await sleep(2)
+                continue
+
+
+def handle_web():
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete()
```

Comment:
You need to pass the runner into this function. Additionally, it's probably best to pass the event loop in.

```python
def handle_web(loop):
    asyncio.set_event_loop(loop)
    loop.run_until_complete(runner())
```

You can pass the loop arg into the args when creating the thread; comment is below.